doc/sphinx/proofs/automatic-tactics/auto.rst

```diff
@@ -310,31 +359,76 @@ pattern or when the hint has no pattern.
-Note that the default transparency
-settings differ between these two methods of creation.  Databases created with
-:cmd:`Create HintDb` have the default setting `Transparent` for both `Variables`
-and `Constants`, while implicitly created databases have the `Opaque` setting.
+The proof search tactics use unification to choose which tactics to try, for example
+whether the goal unifies with a theorem given by :cmd:`Hint Resolve`.
+
+Use :cmd:`Hint Opaque` and :cmd:`Hint Transparent` to control the opacity of
+individual items during this initial unification.  These settings are distinct from the
+non-hint :cmd:`Opaque` and :cmd:`Transparent` settings.  Hint opacity settings
+influence which hints the search tactics try, but in some cases they also affect how
+selected tactics are executed.  (In particular, :tacn:`typeclasses eauto` uses the equivalent of
+:tacn:`autoapply` when applying :cmd:`Hint Resolve` hints, which explicitly use the hint
+opacity settings.  :tacn:`auto` and :tacn:`eauto` don't do this.)
 
 .. cmd:: Create HintDb @ident {? discriminated }
 
-   If there is no hint database named :n:`@name`, creates a new hint database
-   with that name.  Otherwise, does nothing.  The database is
-   implemented by a Discrimination Tree (DT) that serves as a filter to select
-   the lemmas that will be applied. When discriminated, the DT uses
-   transparency information to decide if a constant should considered rigid for
-   filtering, making the retrieval more efficient. By contrast, undiscriminated
-   databases treat all constants as transparent, resulting in a larger
-   number of selected lemmas to be applied, and thus putting more pressure on
-   unification.
+   If there is no hint database named :n:`@ident`, creates a new hint database
+   with that name.  Otherwise, does nothing.
+
+   All constants, variables and projections are set to default to unfoldable (use
+   :cmd:`Hint Constants`, :cmd:`Hint Projections` and :cmd:`Hint Variables`
+   to change this).  For better performance, we recommend setting each of these
+   to :n:`Opaque` and then using :cmd:`Hint Transparent` for specific items
+   when necessary.
+
+   By default, hint databases are undiscriminated.  We recommend using `discriminated`
+   because it generally performs better.
+
+.. _hint_performance_considerations:
+
+Performance considerations
+``````````````````````````
+
+The proof search tactics decide which hints to try by first selecting hints
+whose pattern has the same :term:`head constant` as the goal.  Then the selected
+hints are tried as follows:
+
+- For hints other than :cmd:`Hint Extern`: If the database is discriminated, only
+  matching hints are tried.  Hints match if all compared items that are opaque in
+  both the goal and the pattern are identical.  In this case you can reduce the
+  number of matches by marking definitions as hint opaque through trial and error
+  (if the proof search fails after making something opaque, then it's incorrect).
```

Comment:
```suggestion
  both the goal and the pattern are identical. Marking definitions as Hint Opaque reduces the number of matches, but may cause some hints to no longer match the goal if they previously relied on unfolding of the marked definitions.
```

I don't think trial-and-error performance engineering on a hint database is recommendation-worthy -- for example, it isn't a good idea if there are users of that hintdb that cannot be tried out. Also, I worry that the resulting database would be brittle to small changes. I tried to capture the underlying principle in a more general comment.